kernel/src/hil/kv_store.rs

```diff
@@ -0,0 +1,112 @@
+//! Interface for KV Stores
+
+use crate::returncode::ReturnCode;
+
+#[derive(Clone, Copy, PartialEq)]
+pub enum Operation {
+    None,
+    Init,
+    GetKey,
+    AppendKey,
+    InvalidateKey,
+    GarbageCollect,
+}
+
+/// The type of keys, this should define the output size of the digest
+/// operations.
+pub trait KeyType: Eq + Copy + Clone + Sized + AsRef<[u8]> + AsMut<[u8]> {}
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a, K: KeyType> {
+    /// This callback is called when any operation is complete.
+    ///
+    /// `operation`: The operation that completed
+    /// `result`: Nothing on success, 'ReturnCode' on error
+    /// `key`: A key buffer is one exists
+    /// `value`: A value buffer is one exists
+    /// `buf`: A buf buffer is one exists
+    fn operation_complete(
+        &'a self,
+        operation: Operation,
+        result: Result<(), ReturnCode>,
+        key: Option<&'static mut K>,
+        value: Option<&'static mut [u8]>,
+        buf: Option<&'static mut [u8]>,
+    );
+}
+
+pub trait KVStore {
+    /// The type of the hashed key. For example '[u8; 64]'.
+    type K: KeyType;
+
+    /// Set the client
+    fn set_client(&self, client: & dyn Client<Self::K>);
+
+    /// Appends the key/value pair.
```

Comment:
What happens if the key is already used?

Reply:
That is up to the implementation, maybe some implementations allow that.

Reply:
But what does the user expect to happen? Do they get the last value appended? All of the values appended? Is this how values are updated?

Reply:
They would either get an error or a success value at this call.

When they call `get_value()` I guess that would depend on the implementation.

Reply:
I'm not sure I understand what the purpose of this interface is if how it works is left to a particular implementation. If how a key's value is updated isn't specified then it will be hard for any upper layers to use this, won't it?

Reply:
I don't think so.

An upper layer attempts to add a key. Implementations can decide what they do with that, maybe they support collisions, maybe they don't and return either Success or Error.

Just a note though, that you can't really support collisions when passing in the hashed key, so you pretty much will always have an error. I'll add a function to generate keys to allow custom key types.

Reply:
How does a client update a key with a new value?